### ISSUE TITLE
Help `skipmissing` with a DataFrameRow full of `Missing`s

### ISSUE THREAD

**Issue by pdeffebach** [JuliaData/DataFrames.jl#2164]
In reference to #2161, we have the following behavior. 

```
julia> df = DataFrame(a = [1, missing], b = [2, missing])
2×2 DataFrame
│ Row │ a       │ b       │
│     │ Int64?  │ Int64?  │
├─────┼─────────┼─────────┤
│ 1   │ 1       │ 2       │
│ 2   │ missing │ missing │

julia> sum(skipmissing(df[2, :]))
ERROR: ArgumentError: reducing over an empty collection is not allowed
```

But with NamedTuples we can do 

```
julia> nt = NamedTuple{(:a,:b), Tuple{Union{Int,Missing}, Union{Int,Missing}}}((missing, missing));

julia> sum(skipmissing(nt))
0
```

Given that a DataFrameRow is *kind of* like a NamedTuple we should be able to support this behavior. I don't know off the top of my head what we would need to make this work. Probably a customized `Base.mapreduce_empty` implementation.

**Comment by bkamins** [comment]
The problem is that `DataFrameRow` does not have column types encoded in its type (and this is a typical assumption of `eltype`). The reason is that column types can change at any time (e.g. after you have taken the column type information and before using this information).

Therefore `DataFrameRow` now supports the default and promises `Any` eltype only.

I am not sure what would be a solution here. The problem is that functions in Base mostly assume `eltype` can be determined from type not from value.

**Comment by pdeffebach** [comment]
>  (e.g. after you have taken the column type information and before using this information).

I'm not sure I fully understand. Could you give an example of how this might break? I think `eltype` is all that would be needed here so maybe there is a way to make it work. I will check tomorrow.

**Comment by bkamins** [comment]
```
julia> df = DataFrame(ones(2,3))
2×3 DataFrame
│ Row │ x1      │ x2      │ x3      │
│     │ Float64 │ Float64 │ Float64 │
├─────┼─────────┼─────────┼─────────┤
│ 1   │ 1.0     │ 1.0     │ 1.0     │
│ 2   │ 1.0     │ 1.0     │ 1.0     │

julia> for x in skipmissing(df[1, :])
           df[!, :] .= "a"
           @show x
       end
x = 1.0
x = "a"
x = "a"
```

**Comment by nalimilan** [comment]
I guess we could define `eltype(x::DataFrameRow)  = reduce(promote_type, eltype.(eachcol(parent(x))))`, but that would be quite slow as the compiler would likely not be able to hoist that call out of a loop (especially since that would happen deep in a function like `reduce`).

**Comment by bkamins** [comment]
This would be doable, but very slow as you note, so probably it is better to use `init` in `reduce` as I have recommended anyway.

Still doing this would break the design contract of `eltype` as I have noted. To quote the documentation:

>   eltype(type)
  Determine the type of the elements generated by iterating a collection of
  the given type. For dictionary types, this will be a Pair{KeyType,ValType}.
  The definition eltype(x) = eltype(typeof(x)) is provided for convenience so
  that instances can be passed instead of types. However the form that accepts
  a type argument should be defined for new types.

As you can see - Base is explicit that `eltype` should be a property of type not a value (and as I have said - in Base this assumption is made in the code, i.e. that `eltype` is "immutable" property of a type).

**Comment by pdeffebach** [comment]
`Tables.namedtupleiterator` would make this work. We can point people there if they don't want to modify the data frame in the loop. Perhaps we can also provide a small wrapper for this function as well.

**Comment by bkamins** [comment]
OK - so what would be the recommendation here as an action point? (given we understand how things stay now)

**Comment by nalimilan** [comment]
Maybe recommend `select(df, NT(...) => sum ∘ skipmissing)` (with a better name for `NT`)?

**Comment by bkamins** [comment]
Needs `ByRow` so:
```
select(df, AsTable(...) => ByRow(sum ∘ skipmissing))
```
(with the `AsTable` name, which I currently like :smile:). And we have to remember to put appropriate element types for `NamedTuple` in `ByRow` (maybe it will happen without working on it).

@pdeffebach - where would you want this recommendation to be put?

**Comment by pdeffebach** [comment]
I think `AsTable` is good, but it's not a great name combined with `ByRow`. But I guess a named tuple is still a table in the sense that it looks things up... 

We have 6 months to see what users think of the name. 

With regards to actions we can take right now: this error only pops up when users do something like

```
df.y = map(eachrow(df))
    sum(skipmissing(row))
end
```

Which was very idiomatic julia code until this stuff. I think we should just steer users away from `map(eachrow(df))` and towards `select` and `transform`. They are far better for lots of other reasons too.

**Comment by bkamins** [comment]
Named tuple is a table as it follows Tables.jl interface.

**Comment by pdeffebach** [comment]
Yes, but getting a classroom full of Masters in Public Policy students to understand that `AsTable` returns an object that maps `Symbols` to `Scalars` might be tough. 

Ultimately this is way more clear and intuitive than R, though. And People love Stata's `egen` opaqueness on this. So I think it's worth going with `AsTable`.

**Comment by pdeffebach** [comment]
We could *maybe* export an alias for `Tables.namedtupleiterator`. But maybe not. This would only be when they *don't* want to create a new column in the data frame and rather want to produce a vector. 

But I think we should mirror dplyr here and maybe focus on a workflow that assumes people don't want to produce vectors on their own all the time. If a user is advanced enough to be working with vectors that aren't part of a DataFrame, they can use  `Tables.namedtupleiterator`.

**Comment by bkamins** [comment]
> `AsTable` returns an object that maps `Symbols` to Scalars

No `AsTable` will return a `NamedTuple` of vectors if applied to column selector. Only combined with `ByRow` on a function it will pass `NamedTuple` of scalars to a function.

Essentially we have two orthogonal operations here.

**Comment by pdeffebach** [comment]
Sorry, yes, I should have clarified I was referring to `AsTable([:x2, :x3]) => ByRow(f)`. That's a bit awkward but I think it's fine.

**Comment by bkamins** [comment]
@pdeffebach - can this be closed now?

**Comment by pdeffebach** [comment]
Should we add something in the docs about this? It's a bit niche but maybe someone will come across it.

**Comment by bkamins** [comment]
OK - but where? Adding something like to the docstring (the second sentence is new)?
```
A `DataFrameRow` supports the iteration interface and can therefore be passed to functions that expect a collection as an argument. In such situations `eltype` of `DataFrameRow` is always `Any`.
```

**Comment by pdeffebach** [comment]
Let's discuss in #2196

**Comment by bkamins** [comment]
Fixed by #2196
